### PR TITLE
Update `aggregate_metrics_df` to work with `None` values

### LINKED ISSUE #518: Update `aggregate_metrics_df` to work with `None` values
### 🚀 Task description

If we allow metrics to have `None` value then we should learn to handle these values during aggregation.

The suggested solution is allow aggregation to ignore missing values by creating new types of aggregations like `nanmean` or `nanmedian`.

### Plan

- Edit methods in `METRICS_AGGREGATION_MAP` to handle missing values
- Update `aggregate_metrics_df` to replace NaN with `None` in the end result
  - This could occur if all values are `None`
  - The return signature should be changed in this case
- Clarify in `aggregate_metrics_df` that aggregation by fold is mean and different suffixes only changes the aggregation of segment values

To discuss
- Should we handle case of all `None` values?

### Test cases

- Add tests on `aggregate_metrics_df`
- Add integration test on `Auto` with handling missing values

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by d-a-bunin**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
See #518.

## Closing issues
Closes #518.

### CODE REVIEW

File: etna/metrics/utils.py [hunk 1]
```diff
@@ -37,32 +38,97 @@ def compute_metrics(
...
 ]
 
 METRICS_AGGREGATION_MAP: Dict[MetricAggregationStatistics, Union[str, Callable]] = {
-    "median": "median",
-    "mean": "mean",
-    "std": "std",
+    "median": mean_agg(),
+    "mean": median_agg(),
+    "std": std_agg(),
+    "size": size_agg(),
```
Comment (d-a-bunin) [inline etna/metrics/utils.py:123]: I decided to add this for looking at number of missing values by segments.

File: etna/metrics/utils.py [hunk 2]
```diff
@@ -74,7 +140,7 @@ def aggregate_metrics_df(metrics_df: pd.DataFrame) -> Dict[str, float]:
     if "fold_number" in metrics_df.columns:
         metrics_dict = (
             metrics_df.groupby("segment")
-            .mean()
+            .apply(lambda x: x.mean(skipna=False, numeric_only=True))
```
Comment (d-a-bunin) [inline etna/metrics/utils.py:143]: If we use `.mean()` it works like `nanmean` and ignores folds with missing values. We probably don't want that to retain comparability between segments. 
However, if there is only one value in some fold, the metric can be computed. This situation is very close to not having any value in segment but the result is very different. May be it isn't good.
So, may be we could provide some other methods for aggregation by folds (like nanmean, etc).

Comment (brsnw250) [reply etna/metrics/utils.py:143]: It seems ok to ignore missing values by default when aggregating folds. User agreed to do so when allowed this behavior on the metric level. Sure, it comes at a cost of lacking comparability, but our only alternative to retain full comparability, as I can see right now, is to raise an error on missing values. Which user also can do by not allowing missing values.

File: tests/test_metrics/test_metrics_utils.py [hunk 3]
```diff
@@ -21,3 +26,160 @@ def test_compute_metrics(train_test_dfs: Tuple[TSDataset, TSDataset]):
...
+        "MSE_percentile_5": 3.1,
+        "MSE_percentile_25": 3.5,
+        "MSE_percentile_75": 5.0,
+        "MSE_percentile_95": 5.8,
+    }
+    return result
+
+
+@pytest.fixture
+def metrics_df_with_folds_with_full_missing() -> pd.DataFrame:
```
Comment (brsnw250) [inline tests/test_metrics/test_metrics_utils.py:123]: We should also add a case when whole segment folds are `None`

File: tests/test_auto/conftest.py [hunk 4]
```diff
@@ -35,3 +38,63 @@ class Trial(NamedTuple):
...
+
+    df = pd.concat([df1, df2]).reset_index(drop=True)
+    df = TSDataset.to_dataset(df)
+    tsds = TSDataset(df, freq="D")
+
+    return tsds
+
+
+@pytest.fixture
+def ts_with_few_missing(random_seed) -> TSDataset:
```
Comment (brsnw250) [inline tests/test_auto/conftest.py:84]: We also need with full missing horizon

**Review by brsnw250 — CHANGES_REQUESTED** [review]

File: tests/test_auto/conftest.py (hunk 4 above)
Comment (d-a-bunin) [reply tests/test_auto/conftest.py:84]: Two previous fixtures have 7 elements missing, it was enough to make a whole fold to become missing.

File: tests/test_metrics/test_metrics_utils.py (hunk 3 above)
Comment (d-a-bunin) [reply tests/test_metrics/test_metrics_utils.py:123]: Do you mean for all folds to be filled with None?

File: etna/metrics/utils.py [hunk 5]
```diff
@@ -37,32 +38,97 @@ def compute_metrics(
...
+                message="Degrees of freedom <= 0",
+                action="ignore",
+            )
+            return np.nanstd(a=x.values)
+
+    func.__name__ = "std"
+    return func
+
+
+def size_agg():
```
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/metrics/utils.py:89]: This name is probably too large.
May be we could explain existing names somewhere in `aggregate_metrics_df` or in some other place.

File: tests/test_auto/test_auto.py [hunk 6]
```diff
@@ -72,6 +90,40 @@ def test_objective(
...
+        ts=ts,
+        target_metric=target_metric,
+        metric_aggregation=metric_aggregation,
+        metrics=metrics,
+        backtest_params=backtest_params,
+        initializer=initializer,
+        callback=callback,
+    )
+
+    # TODO: discuss the error here
```
Comment (d-a-bunin) [inline tests/test_auto/test_auto.py:122]: We should discuss this: it fails before we go into metrics computation.

File: tests/test_auto/test_tune.py [hunk 7]
```diff
@@ -53,6 +65,38 @@ def test_objective(
...
+        params_to_tune=params_to_tune,
+        target_metric=target_metric,
+        metric_aggregation=metric_aggregation,
+        metrics=metrics,
+        backtest_params=backtest_params,
+        initializer=initializer,
+        callback=callback,
+    )
+
+    # TODO: discuss the error here
```
Comment (d-a-bunin) [inline tests/test_auto/test_tune.py:95]: We should discuss this: it fails before we go into metrics computation.

File: tests/test_metrics/test_metrics_utils.py [hunk 8]
```diff
@@ -172,10 +103,6 @@ def aggregated_metrics_df_with_full_missing() -> Dict[str, Any]:
     [
         ("metrics_df_with_folds", "aggregated_metrics_df"),
         ("metrics_df_no_folds", "aggregated_metrics_df"),
-        ("metrics_df_with_folds_with_missing", "aggregated_metrics_df_with_missing"),
-        ("metrics_df_no_folds_with_missing", "aggregated_metrics_df_with_missing"),
-        ("metrics_df_with_folds_with_full_missing", "aggregated_metrics_df_with_full_missing"),
-        ("metrics_df_no_folds_with_full_missing", "aggregated_metrics_df_with_full_missing"),
```
Comment (brsnw250) [inline tests/test_metrics/test_metrics_utils.py]: Why did you remove these tests?

File: tests/test_auto/test_auto.py (hunk 6 above)
Comment (brsnw250) [reply tests/test_auto/test_auto.py:122]: Seems like we should support this behavior now. We can leave it as is for this task and create a separate issue to change this. What do you think?

Comment (brsnw250) [reply tests/test_auto/test_auto.py:122]: Or we can try to mock here the part that blocks us from testing intended behavior.

File: tests/test_metrics/test_metrics_utils.py (hunk 8 above)
Comment (d-a-bunin) [reply tests/test_metrics/test_metrics_utils.py]: I just rearranged the tests. I put all of the possible cases in a single fixture.

**Review by brsnw250 — DISMISSED** [review]

**Review by brsnw250 — APPROVED** [review]
